requestmanager/client.go

```diff
@@ -0,0 +1,359 @@
+	terminated := make(chan error, 1)
+	rm.send(&cancelRequestMessage{requestID, false, terminated, graphsync.RequestClientCancelledErr{}}, ctx.Done())
+	select {
+	case <-rm.ctx.Done():
+		return errors.New("context cancelled")
+	case err := <-terminated:
+		return err
+	}
+}
+
+// ProcessResponses ingests the given responses from the network and
+// and updates the in progress requests based on those responses.
+func (rm *RequestManager) ProcessResponses(p peer.ID, responses []gsmsg.GraphSyncResponse,
+	blks []blocks.Block) {
+	rm.send(&processResponseMessage{p, responses, blks}, nil)
+}
+
+// UnpauseRequest unpauses a request that was paused in a block hook based request ID
+// Can also send extensions with unpause
+func (rm *RequestManager) UnpauseRequest(requestID graphsync.RequestID, extensions ...graphsync.ExtensionData) error {
+	response := make(chan error, 1)
+	rm.send(&unpauseRequestMessage{requestID, extensions, response}, nil)
+	select {
+	case <-rm.ctx.Done():
+		return errors.New("context cancelled")
+	case err := <-response:
+		return err
+	}
+}
+
+// PauseRequest pauses an in progress request (may take 1 or more blocks to process)
+func (rm *RequestManager) PauseRequest(requestID graphsync.RequestID) error {
+	response := make(chan error, 1)
+	rm.send(&pauseRequestMessage{requestID, response}, nil)
+	select {
+	case <-rm.ctx.Done():
+		return errors.New("context cancelled")
+	case err := <-response:
+		return err
+	}
+}
+
+// ProcessBlockHooks processes block hooks for the given response & block and cancels
+// the request as needed
+func (rm *RequestManager) ProcessBlockHooks(p peer.ID, response graphsync.ResponseData, block graphsync.BlockData) error {
+	result := rm.blockHooks.ProcessBlockHooks(p, response, block)
+	if len(result.Extensions) > 0 {
+		updateRequest := gsmsg.UpdateRequest(response.RequestID(), result.Extensions...)
+		rm.SendRequest(p, updateRequest)
+	}
+	if result.Err != nil {
+		_, isPause := result.Err.(hooks.ErrPaused)
+		rm.send(&cancelRequestMessage{response.RequestID(), isPause, nil, nil}, nil)
+	}
+	return result.Err
+}
+
+// TerminateRequest marks a request done
+func (rm *RequestManager) TerminateRequest(requestID graphsync.RequestID) {
+	rm.send(&terminateRequestMessage{requestID}, nil)
+}
+
+// SendRequest sends a request to the message queue
+func (rm *RequestManager) SendRequest(p peer.ID, request gsmsg.GraphSyncRequest) {
+	sub := notifications.NewTopicDataSubscriber(&reqSubscriber{p, request, rm.networkErrorListeners})
+	failNotifee := notifications.Notifee{Data: requestNetworkError, Subscriber: sub}
+	rm.peerHandler.AllocateAndBuildMessage(p, 0, func(builder *gsmsg.Builder) {
+		builder.AddRequest(request)
+	}, []notifications.Notifee{failNotifee})
+}
+
+// Startup starts processing for the WantManager.
+func (rm *RequestManager) Startup() {
+	go rm.run()
+}
+
+// Shutdown ends processing for the want manager.
+func (rm *RequestManager) Shutdown() {
+	rm.cancel()
+}
```

Comment:
I'm a bit confused; if the user supplies a context when creating a RequestManager, could they not just cancel the context themselves to do a shutdown? Having RequestManager hold the cancel func and expose a Shutdown method feels unnecessary.

One reason we might have for exposing Shutdown methods is to *also* block until all goroutines have stopped and resources have been freed up. If we think we'll want that later on, then I think leaving these in with a TODO makes sense. Otherwise, I'd personally remove the Shutdown APIs.

Reply:
Also, if we went with the "we'll eventually want to block until all request goroutines are stopped" route, then I think these methods should gain a context parameter, too. That way, I can say "graceful shutdown for 10s to stop accepting requests, and after that, kill all requests which aren't done yet". This is how net/http does it: https://pkg.go.dev/net/http#Server.Shutdown

Reply:
I agree this is odd. This is the pattern of the repo. I'd rather revisit the shutdown behavior globally in a seperate ticket.